crates/core/tedge/src/cli/certificate/create_key.rs

```diff
@@ -0,0 +1,283 @@
+use tedge_config::TEdgeConfig;
+use tedge_p11_server::pkcs11::CreateKeyParams;
+use tedge_p11_server::pkcs11::KeyTypeParams;
+use tedge_p11_server::CryptokiConfig;
+
+use crate::cli::common::Cloud;
+use crate::command::Command;
+use crate::log::MaybeFancy;
+
+pub struct CreateKeyPkcs11Cmd {
+    pub cryptoki_config: CryptokiConfig,
+    pub bits: RsaBits,
+    pub curve: EcCurve,
+    pub label: String,
+    pub r#type: KeyType,
+    pub id: Option<String>,
+    pub cloud: Option<Cloud>,
+    pub token: Option<String>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, ValueEnum)]
+pub enum KeyType {
+    Rsa,
+    Ecdsa,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, ValueEnum)]
+pub enum RsaBits {
+    #[value(name = "2048")]
+    Bits2048,
+    #[value(name = "3072")]
+    Bits3072,
+    #[value(name = "4096")]
+    Bits4096,
+}
+
+impl From<RsaBits> for u16 {
+    fn from(value: RsaBits) -> Self {
+        match value {
+            RsaBits::Bits2048 => 2048,
+            RsaBits::Bits3072 => 3072,
+            RsaBits::Bits4096 => 4096,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, ValueEnum)]
+pub enum EcCurve {
+    P256,
+    P384,
+}
+
+impl From<EcCurve> for u16 {
+    fn from(value: EcCurve) -> Self {
+        match value {
+            EcCurve::P256 => 256,
+            EcCurve::P384 => 384,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Command for CreateKeyPkcs11Cmd {
+    fn description(&self) -> String {
+        "Generate a keypair.".into()
+    }
+
+    async fn execute(&self, config: TEdgeConfig) -> Result<(), MaybeFancy<anyhow::Error>> {
+        let key = match self.r#type {
+            KeyType::Rsa => KeyTypeParams::Rsa {
+                bits: self.bits.into(),
+            },
+            KeyType::Ecdsa => KeyTypeParams::Ec {
+                curve: self.curve.into(),
+            },
+        };
+        let id = self
+            .id
+            .as_ref()
+            .map(|s| parse_id(s))
+            .transpose()
+            .context("invalid id")?;
+
+        let cryptoki = tedge_p11_server::tedge_p11_service(self.cryptoki_config.clone())?;
+        let Some(token) = self.token.clone() else {
+            eprintln!("No token URL was provided for this operation; the available tokens are:");
+            let tokens = cryptoki.get_tokens_uris()?;
+            for token_uri in tokens {
+                eprintln!("{token_uri}");
+            }
+            return Ok(());
+        };
+
+        let params = CreateKeyParams {
+            key,
+            label: self.label.clone(),
+            id,
+        };
+
+        // generate a keypair
+        // should probably verify the keys before using them
+        let key = cryptoki.create_key(&token, params)?;
+        let pubkey_pem = key.pem;
+        let uri = key.uri;
+
+        eprintln!("New keypair was successfully created.");
+        eprintln!("Key URI: {uri}");
+        eprintln!("Public key:\n{pubkey_pem}\n");
```

Comment:
I'm wondering if adding an option to save the public key to file would make more sense, as it would make it more compatible with other 3rd party tooling  (e.g. gnutls's certtool).

Something like `--out-public-key` (though happy with a better name)?

And the user could provide `--out-public-key -` to denote writing the output to stdout.